services/telemetry/config.go

```diff
@@ -0,0 +1,313 @@
+package telemetry
+
+import (
+	_ "embed" //nolint:golint
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"aead.dev/minisign"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v3"
+)
+
+// ServiceConfig telemetry config.
+type ServiceConfig struct {
+	l                          *logrus.Entry
+	Enabled                    bool   `yaml:"enabled"`
+	LoadDefaults               bool   `yaml:"load_defaults"`                //nolint:tagliatelle
+	ConfigLocation             string `yaml:"config_location"`              //nolint:tagliatelle
+	DisableSigningVerification bool   `yaml:"disable_signing_verification"` //nolint:tagliatelle
+	Signing                    struct {
+		TrustedPublicKeys       []string             `yaml:"trusted_public_keys"` //nolint:tagliatelle
+		trustedPublicKeysParsed []minisign.PublicKey `yaml:"-"`
+	} `yaml:"signing"`
+	telemetry    []Config        `yaml:"-"`
+	Endpoints    EndpointsConfig `yaml:"endpoints"`
+	SaasHostname string          `yaml:"saas_hostname"` //nolint:tagliatelle
+	DataSources  struct {
+		VM          *DataSourceVictoriaMetrics `yaml:"VM"`
+		QanDBSelect *DSConfigQAN               `yaml:"QANDB_SELECT"` //nolint:tagliatelle
+		PmmDBSelect *DSConfigPMMDB             `yaml:"PMMDB_SELECT"` //nolint:tagliatelle
+	} `yaml:"datasources"`
+	Reporting ReportingConfig `yaml:"reporting"`
+}
+
+// FileConfig top level telemetry config element.
+type FileConfig struct {
+	Telemetry []Config `yaml:"telemetry"`
+}
+
+// EndpointsConfig telemetry endpoint config.
+type EndpointsConfig struct {
+	Report string `yaml:"report"`
+}
+
+// ReportEndpointURL returns reporting endpoint URL.
+func (c *ServiceConfig) ReportEndpointURL() string {
+	return fmt.Sprintf(c.Endpoints.Report, c.SaasHostname)
+}
+
+// DSConfigQAN telemetry config.
+type DSConfigQAN struct {
+	Enabled    bool          `yaml:"enabled"`
+	Timeout    time.Duration `yaml:"-"`
+	TimeoutStr string        `yaml:"timeout"`
+	DSN        string        `yaml:"dsn"`
+}
+
+// DataSourceVictoriaMetrics telemetry config.
+type DataSourceVictoriaMetrics struct {
+	Enabled    bool          `yaml:"enabled"`
+	Timeout    time.Duration `yaml:"-"`
+	TimeoutStr string        `yaml:"timeout"`
+	Address    string        `yaml:"address"`
+}
+
+// DSConfigPMMDB telemetry config.
+type DSConfigPMMDB struct {
+	Enabled                bool          `yaml:"enabled"`
+	Timeout                time.Duration `yaml:"-"`
+	TimeoutStr             string        `yaml:"timeout"`
+	UseSeparateCredentials bool          `yaml:"use_separate_credentials"` //nolint:tagliatelle
+	// Credentials used by PMM
+	DSN struct {
+		Scheme string
+		Host   string
+		DB     string
+		Params string
+	} `yaml:"-"`
+	Credentials struct {
+		Username string
+		Password string
+	} `yaml:"-"`
+	SeparateCredentials struct {
+		Username string `yaml:"username"`
+		Password string `yaml:"password"`
+	} `yaml:"separate_credentials"` //nolint:tagliatelle
```

Comment:
we can get use only `Credentials` instead of two separate fields, we can check for `UseSeparateCredentials` in main() where we fill values for Credentials

Reply:
we need it to be able to provide custom credentials